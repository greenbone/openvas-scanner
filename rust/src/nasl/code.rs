use std::path::{Path, PathBuf};

use codespan_reporting::files::SimpleFile;

use super::{
    Loader,
    error::Level,
    syntax::{DescriptionBlock, LoadError, ParseError, Parser, Tokenizer, grammar::Ast},
};

fn parse(code: &str) -> Result<Ast, Vec<ParseError>> {
    let tokenizer = Tokenizer::tokenize(code);
    let parser = Parser::new(tokenizer);
    parser.parse_program()
}

fn parse_description_block(code: &str) -> Result<Ast, Vec<ParseError>> {
    let tokenizer = Tokenizer::tokenize(code);
    let mut parser = Parser::new(tokenizer);
    let result: Result<DescriptionBlock, ParseError> =
        parser.parse_span().map_err(|e| e.unwrap_as_spanned());
    result
        .map(|metadata| metadata.into_ast())
        .map_err(|e| vec![e])
}

pub type SourceFile = SimpleFile<String, String>;

pub struct ParseResult {
    result: Result<Ast, Vec<ParseError>>,
    file: SourceFile,
}

impl ParseResult {
    pub fn new(code: &str, path: &Path) -> Self {
        Self::new_with_parse_fn(code, path, parse)
    }

    pub fn new_with_parse_fn(
        code: &str,
        path: &Path,
        f: impl Fn(&str) -> Result<Ast, Vec<ParseError>>,
    ) -> Self {
        let file = SimpleFile::new(path.to_string_lossy().into(), code.to_owned());
        let result = f(code);
        Self { file, result }
    }

    pub fn result(self) -> Result<Ast, Vec<ParseError>> {
        self.result
    }

    pub fn num_errors(&self) -> usize {
        match &self.result {
            Ok(_) => 0,
            Err(errs) => errs.len(),
        }
    }

    pub fn emit_errors(self) -> Result<Ast, Vec<ParseError>> {
        match self.result {
            Ok(result) => Ok(result),
            Err(errors) => {
                super::error::emit_errors(&self.file, errors.iter().cloned(), Level::Error);
                Err(errors)
            }
        }
    }

    pub fn emit_errors_get_ast_and_file(self) -> Result<(Ast, SourceFile), Vec<ParseError>> {
        match self.result {
            Ok(result) => Ok((result, self.file)),
            Err(errors) => {
                super::error::emit_errors(&self.file, errors.iter().cloned(), Level::Error);
                Err(errors)
            }
        }
    }

    #[cfg(test)]
    pub fn unwrap_errors_str(self) -> String {
        let errs = self.result.unwrap_err();
        super::error::emit_errors_str(&self.file, errs.into_iter())
    }

    pub fn file(&self) -> &SourceFile {
        &self.file
    }
}

pub struct Code {
    code: String,
    path: Option<PathBuf>,
}

impl Code {
    pub fn load(loader: &dyn Loader, path: impl AsRef<Path>) -> Result<Self, LoadError> {
        Ok(Self {
            code: loader.load(&path.as_ref().to_string_lossy())?,
            path: Some(path.as_ref().to_owned()),
        })
    }

    pub fn from_string(code: &str) -> Self {
        Self {
            path: None,
            code: code.to_string(),
        }
    }

    pub fn from_string_filename(code: &str, filename: impl AsRef<Path>) -> Self {
        Self {
            code: code.to_string(),
            path: Some(filename.as_ref().to_owned()),
        }
    }

    pub fn source_file(&self) -> SourceFile {
        let path_str = self
            .path
            .as_ref()
            .map(|path| path.to_string_lossy().to_string())
            .unwrap_or("".to_string());
        SimpleFile::new(path_str, self.code().to_owned())
    }

    #[cfg(test)]
    pub fn tokenize(self) -> TokenizeResult {
        match self.path {
            Some(path) => TokenizeResult::new(&self.code, &path),
            None => unimplemented!(),
        }
    }

    pub fn parse(self) -> ParseResult {
        match self.path {
            Some(path) => ParseResult::new(&self.code, &path),
            None => ParseResult::new(&self.code, Path::new("")),
        }
    }

    pub fn parse_description_block(self) -> ParseResult {
        let path = self.path.unwrap_or(Path::new("").to_owned());
        ParseResult::new_with_parse_fn(&self.code, &path, parse_description_block)
    }

    pub fn code(&self) -> &str {
        &self.code
    }
}

#[cfg(test)]
pub use tokenize::TokenizeResult;

#[cfg(test)]
mod tokenize {
    use std::path::Path;

    use codespan_reporting::files::SimpleFile;
    use itertools::{Either, Itertools};

    use crate::nasl::{
        error::Level,
        syntax::{Token, Tokenizer, TokenizerError},
    };

    use super::{super::error, SourceFile};

    fn split_tokens_and_errors(tokenizer: Tokenizer) -> Result<Vec<Token>, Vec<TokenizerError>> {
        let (tokens, errors): (Vec<_>, Vec<_>) = tokenizer.partition_map(|a| match a {
            Ok(a) => Either::Left(a),
            Err(a) => Either::Right(a),
        });
        if !errors.is_empty() {
            Err(errors)
        } else {
            Ok(tokens)
        }
    }

    pub struct TokenizeResult {
        result: Result<Vec<Token>, Vec<TokenizerError>>,
        file: SourceFile,
    }

    impl TokenizeResult {
        pub fn new(code: &str, path: &Path) -> Self {
            let file = SimpleFile::new(path.to_string_lossy().into(), code.to_owned());
            let result = split_tokens_and_errors(Tokenizer::tokenize(code));
            Self { file, result }
        }

        pub fn result(self) -> Result<Vec<Token>, Vec<TokenizerError>> {
            self.result
        }

        pub fn emit_errors(self) -> Option<Vec<Token>> {
            match self.result {
                Ok(result) => Some(result),
                Err(errors) => {
                    error::emit_errors(&self.file, errors.into_iter(), Level::Error);
                    None
                }
            }
        }

        pub fn unwrap_errors_str(self) -> String {
            let errs = self.result.unwrap_err();
            error::emit_errors_str(&self.file, errs.into_iter())
        }
    }
}
