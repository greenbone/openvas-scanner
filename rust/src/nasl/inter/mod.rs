mod assign;
mod call;
mod declare;
mod forking_interpreter;
mod include;
mod interpreter;
mod loop_extension;
mod operator;

pub use forking_interpreter::ForkingInterpreter;
