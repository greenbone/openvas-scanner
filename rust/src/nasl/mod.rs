mod builtin;
pub mod interpreter;
pub mod syntax;
pub mod utils;

#[cfg(test)]
pub mod test_utils;

pub mod prelude {
    pub use super::builtin::ContextFactory;
    pub use super::builtin::RegisterBuilder;
    pub use super::syntax::FSPluginLoader;
    pub use super::syntax::Loader;
    pub use super::syntax::NaslValue;
    pub use super::utils::function::CheckedPositionals;
    pub use super::utils::function::FromNaslValue;
    pub use super::utils::function::Positionals;
    pub use super::utils::function::ToNaslResult;
    pub use super::utils::Context;
    pub use super::utils::ContextType;
    pub use super::utils::FunctionErrorKind;
    pub use super::utils::NaslResult;
    pub use super::utils::Register;
    pub use crate::function_set;
    pub use nasl_function_proc_macro::nasl_function;
}

pub use prelude::*;

#[cfg(test)]
pub mod test_prelude {
    pub use super::prelude::*;
    pub use super::test_utils::check_code_result;
    pub use super::test_utils::DefaultTestBuilder;
    pub use super::test_utils::TestBuilder;
    pub use crate::check_code_result_matches;
    pub use crate::check_err_matches;
}

pub use builtin::nasl_std_functions;

pub use syntax::NoOpLoader;
