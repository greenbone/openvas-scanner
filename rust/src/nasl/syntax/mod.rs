// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]
pub mod grammar;
mod loader;
mod metadata;
mod parser;
mod token;
mod tokenizer;

pub use loader::*;
pub use metadata::DescriptionBlock;
pub use parser::{ParseError, Parser};
pub(super) use token::Ident;
pub use token::Keyword;
pub(super) use token::LiteralKind;
pub use token::Token;
pub use token::TokenKind;
pub use tokenizer::CharIndex;
pub use tokenizer::Tokenizer;
pub use tokenizer::TokenizerError;
