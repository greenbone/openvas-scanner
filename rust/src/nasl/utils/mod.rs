// SPDX-FileCopyrightText: 2024 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]
pub mod error;
mod executor;
pub mod function;
pub mod hosts;
pub mod lookup_keys;
pub mod scan_ctx;

pub use super::interpreter::Register;
pub use error::ArgumentError;
pub use error::FnError;
pub use error::InternalError;
pub use scan_ctx::{ScanCtx, ScriptCtx};

pub use executor::{Executor, IntoFunctionSet, NaslFunction, StoredFunctionSet};

use crate::nasl::interpreter::NaslValue;

/// The result of a function call.
pub type NaslResult = Result<NaslValue, FnError>;

/// Resolves positional arguments from the register.
pub fn resolve_positional_arguments(register: &Register) -> Vec<NaslValue> {
    match register.nasl_value(lookup_keys::FC_ANON_ARGS) {
        Ok(NaslValue::Array(arr)) => arr.to_vec(),
        Ok(unexpected) => {
            tracing::warn!(
                "expected array but got: {:?}. Maybe {} was overridden. Ignoring.",
                unexpected,
                lookup_keys::FC_ANON_ARGS
            );
            vec![]
        }
        Err(_) => vec![],
    }
}

/// Allows the definition of global variables
/// belonging to certain builtin functions.
// This is currently only used with `experimental` feature,
// so we mark it as public in order to save many feature gates.
pub trait DefineGlobalVars {
    fn get_global_vars() -> Vec<(&'static str, NaslValue)>;
}
