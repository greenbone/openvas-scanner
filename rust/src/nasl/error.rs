use std::ops::Range;

use codespan_reporting::diagnostic::Diagnostic;
use codespan_reporting::diagnostic::Label;
use codespan_reporting::term;
use codespan_reporting::term::termcolor::Buffer;
use codespan_reporting::term::termcolor::ColorChoice;
use codespan_reporting::term::termcolor::StandardStream;

use super::code::SourceFile;
use super::syntax::CharIndex;

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct Span {
    start: CharIndex,
    end: CharIndex,
}

impl Span {
    pub(crate) fn new(start: CharIndex, end: CharIndex) -> Self {
        assert!(start.0 <= end.0);
        Self { start, end }
    }

    pub(crate) fn start(&self) -> CharIndex {
        self.start
    }

    pub(crate) fn end(&self) -> CharIndex {
        self.end
    }

    pub(crate) fn join(&self, span: Span) -> Span {
        Span::new(
            CharIndex(self.start.0.min(span.start.0)),
            CharIndex(self.end.0.max(span.end.0)),
        )
    }
}

impl From<Span> for Range<usize> {
    fn from(value: Span) -> Range<usize> {
        value.start.0..value.end.0
    }
}

pub trait Spanned {
    fn span(&self) -> Span;
}

impl Spanned for Span {
    fn span(&self) -> Span {
        *self
    }
}

pub trait AsCodespanError {
    fn span(&self) -> Span;
    fn message(&self) -> String;
}

pub enum Level {
    Warn,
    Error,
}

pub fn emit_errors<T: AsCodespanError>(
    file: &SourceFile,
    errs: impl Iterator<Item = T>,
    level: Level,
) {
    let writer = StandardStream::stderr(ColorChoice::Always);
    let config = codespan_reporting::term::Config::default();
    for err in errs {
        let diagnostic = match level {
            Level::Warn => Diagnostic::warning(),
            Level::Error => Diagnostic::error(),
        };
        let diagnostic = diagnostic.with_message(err.message()).with_labels(vec![
            Label::primary((), err.span()).with_message(err.message()),
        ]);
        term::emit(&mut writer.lock(), &config, file, &diagnostic).unwrap();
    }
}

pub fn emit_errors_str<T: AsCodespanError>(
    file: &SourceFile,
    errs: impl Iterator<Item = T>,
) -> String {
    let mut writer = Buffer::no_color();
    let config = codespan_reporting::term::Config::default();
    for err in errs {
        let diagnostic = Diagnostic::error()
            .with_message(err.message())
            .with_labels(vec![
                Label::primary((), err.span()).with_message(err.message()),
            ]);
        term::emit(&mut writer, &config, file, &diagnostic).unwrap();
    }
    String::from_utf8(writer.into_inner()).unwrap()
}
