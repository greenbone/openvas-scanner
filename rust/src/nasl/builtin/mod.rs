// SPDX-FileCopyrightText: 2024 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]

mod array;
mod cert;
mod cryptographic;
mod description;
mod error;
mod host;
mod http;
mod isotime;
mod knowledge_base;
mod misc;
mod network;
mod preferences;
#[cfg(feature = "nasl-builtin-raw-ip")]
pub mod raw_ip;
mod regex;
mod report_functions;
mod ssh;
mod string;
mod sys;

#[cfg(test)]
mod tests;

pub use error::BuiltinError;
pub use host::HostError;
pub use knowledge_base::KBError;

use super::utils::Executor;

pub use network::socket::NaslSockets;

/// Creates a new Executor and adds all the functions to it.
pub fn nasl_std_functions() -> Executor {
    let mut executor = Executor::default();
    executor
        .add_set(array::Array)
        .add_set(report_functions::Reporting::default())
        .add_set(knowledge_base::KnowledgeBase)
        .add_set(misc::Misc)
        .add_set(string::NaslString)
        .add_set(host::Host)
        .add_set(http::NaslHttp2::default())
        .add_set(http::NaslHttp)
        .add_set(network::socket::SocketFns)
        .add_set(network::network::Network)
        .add_set(regex::RegularExpressions)
        .add_set(cryptographic::Cryptographic)
        .add_set(description::Description)
        .add_set(isotime::NaslIsotime)
        .add_set(preferences::Preferences)
        .add_set(cryptographic::rc4::CipherHandlers::default())
        .add_set(sys::Sys)
        .add_set(ssh::Ssh::default())
        .add_set(cert::NaslCerts::default());

    #[cfg(feature = "nasl-builtin-raw-ip")]
    executor.add_set(raw_ip::RawIp);
    #[cfg(feature = "nasl-builtin-raw-ip")]
    executor.add_global_vars(raw_ip::RawIp);

    executor
}
