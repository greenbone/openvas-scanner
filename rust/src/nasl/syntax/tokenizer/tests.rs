use crate::nasl::Code;

use super::*;

pub fn tokenize_ok(file_name: &str, code: &str) -> Vec<Token> {
    let results = Code::from_string_filename(code, file_name).tokenize();
    results.emit_errors().unwrap()
}

pub fn tokenize_err(file_name: &str, code: &str) -> String {
    let results = Code::from_string_filename(code, file_name).tokenize();
    results.unwrap_errors_str()
}

macro_rules! test_ok {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_debug_snapshot!(tokenize_ok(stringify!($name), $code));
        }
    };
}

macro_rules! test_err {
    ($name: ident, $code: literal) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!(tokenize_err(stringify!($name), $code));
        }
    };
}

test_ok!(skip_whitespace, "     (       ");

test_ok!(one_char_tokens, "( ) [ ] { } . - + % ; / * : ~ & | ^");
test_ok!(
    two_char_tokens,
    "& && | || ! != !~ = == =~ > >> >= >< < << <= - -- + += ++ / /= * ** *="
);
test_ok!(three_char_tokens, ">>> >>= >!< <<=");

test_ok!(four_symbol_tokens, ">>>=");

test_ok!(string, "\"I am a closed string \"");
test_err!(unclosed_string, "\"hello I am a unclosed string\\");

test_ok!(data_string, r#"'Hello \\\'you\\\'!'"#);
test_err!(unclosed_data_string, "'Hello \\'you\\'!\\'");

test_ok!(simple_number, "1");
test_ok!(
    numbers,
    "0 0b01 1234567890 012345670 0x1234567890ABCDEF 0b02"
);
test_err!(invalid_numbers, "0x 0b 0b2");
test_err!(invalid_hex_characters, "0x123h");

test_ok!(single_line_comments, "# this is a comment\n;");

test_ok!(identifier, "help_lo _hello _h4llo 4_h4llo");

test_ok!(
    keywords,
    "for foreach if else while repeat until local_var global_var return include exit break continue"
);
test_ok!(keyword_literals, "TRUE FALSE NULL");

test_ok!(string_quoting, r"'webapps\\\\appliance\\\\'");

test_ok!(data_escape_quoting, r#"'version=\"1.0\"'"#);

test_ok!(simplified_ipv4_address, "10.187.76.12");
test_err!(wrong_ipv4_address, "10.0x 10.0.x 10.0.0.x");

test_ok!(repeat_x_times, "x() x 10;");

test_ok!(use_tokenizer, "local_var hello = 'World!';");
