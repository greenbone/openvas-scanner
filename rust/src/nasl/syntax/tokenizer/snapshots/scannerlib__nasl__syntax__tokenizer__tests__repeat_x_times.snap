---
source: src/nasl/syntax/tokenizer/tests.rs
expression: "tokenize_ok(stringify! (repeat_x_times), \"x() x 10;\")"
---
[
    Token {
        kind: Ident(
            Ident {
                ident: "x",
                span: Span {
                    start: CharIndex(
                        0,
                    ),
                    end: CharIndex(
                        1,
                    ),
                },
            },
        ),
        span: Span {
            start: CharIndex(
                0,
            ),
            end: CharIndex(
                1,
            ),
        },
    },
    Token {
        kind: LeftParen,
        span: Span {
            start: CharIndex(
                1,
            ),
            end: CharIndex(
                2,
            ),
        },
    },
    Token {
        kind: RightParen,
        span: Span {
            start: CharIndex(
                2,
            ),
            end: CharIndex(
                3,
            ),
        },
    },
    Token {
        kind: X,
        span: Span {
            start: CharIndex(
                4,
            ),
            end: CharIndex(
                6,
            ),
        },
    },
    Token {
        kind: Literal(
            Literal {
                kind: Number(
                    10,
                ),
                span: Span {
                    start: CharIndex(
                        6,
                    ),
                    end: CharIndex(
                        8,
                    ),
                },
            },
        ),
        span: Span {
            start: CharIndex(
                6,
            ),
            end: CharIndex(
                8,
            ),
        },
    },
    Token {
        kind: Semicolon,
        span: Span {
            start: CharIndex(
                8,
            ),
            end: CharIndex(
                9,
            ),
        },
    },
]
